include/seqan3/io/alignment_file/format_sam.hpp

```diff
@@ -0,0 +1,571 @@
+{
+public:
+    /*!\name Constructors, destructor and assignment
+     * \brief Rule of five explicitly defaulted.
+     * \{
+     */
+    alignment_file_format_sam() = default;
+    alignment_file_format_sam(alignment_file_format_sam const &) = delete;
+    alignment_file_format_sam & operator=(alignment_file_format_sam const &) = delete;
+    alignment_file_format_sam(alignment_file_format_sam &&) = default;
+    alignment_file_format_sam & operator=(alignment_file_format_sam &&) = default;
+    //!\}
+
+    //!\brief The valid file extensions for this format; note that you can modify this value.
+    static inline std::vector<std::string> file_extensions
+    {
+        { "sam" },
+    };
+
+    //!\copydoc alignment_file_output_format_concept::write
+    template <typename stream_type,
+              typename header_type,
+              typename seq_type,
+              typename id_type,
+              typename offset_type,
+              typename ref_seq_type,
+              typename ref_id_type,
+              typename ref_offset_type,
+              typename align_type,
+              typename flag_type,
+              typename mapq_type,
+              typename qual_type,
+              typename mate_type,
+              typename tag_dict_type,
+              typename evalue_type,
+              typename bit_score_type>
+    void write(stream_type                      &  stream,
+               alignment_file_out_options const &  options,
+               header_type                      && header,
+               seq_type                         && seq,
+               id_type                          && id,
+               offset_type                      && offset,
+               ref_seq_type                     && SEQAN3_DOXYGEN_ONLY(ref_seq),
+               ref_id_type                      && ref_id,
+               ref_offset_type                  && ref_offset,
+               align_type                       && align,
+               flag_type                        && flag,
+               mapq_type                        && mapq,
+               qual_type                        && qual,
+               mate_type                        && mate,
+               tag_dict_type                    && tag_dict,
+               evalue_type                      && SEQAN3_DOXYGEN_ONLY(evalue),
+               bit_score_type                   && SEQAN3_DOXYGEN_ONLY(bit_score))
+    {
+        /* Note the following general things:
+         *
+         * - Given the SAM specifications, all fields may be empty
+         *
+         * - Arithmetic values default to 0 while all others default to '*'
+         *
+         * - Because of the former, arithmetic values can be directly streamed
+         *   into 'stream' as operator<< is defined for all arithmetic types
+         *   and the default value (0) is also the SAM default.
+         *
+         * - All other non-arithmetic values need to be checked for emptiness
```

Comment:
do you mean that the file class should pass default initialized fields?

Reply:
Yes, if the behaviour for `std::ignore` and "" is exactly the same and if there are no illegal combinations of unselected fields that could/should be detected at compile-time, we might as well get rid of the std::ignores!

Reply:
Unfortunately it is not.. 
1) If the alignment type does not decay to ignore, the seq type and offset type mustn't either
2) If neither the header type nor the ref_id type decay to ignore, the ref_id must be checked against the ref_dict

In addition, I think default values are not possible since they differ between BLAST and SAM. The BLAST format will just not print the column whereas SAM print default values.